---
title: Modelling yeast cross from two parents
authors:
  - name: Markus Owen
format: 
  html:
    embed-resources: true
    code-fold: true
    number-sections: true
    toc: true
    toc-depth: 3
    date: now
    date-modified: last-modified
    date-format: "MMMM DD, YYYY, HH:mm:ss"
  pdf: 
    number-sections: true
    colorlinks: true
execute:
  echo: true
jupyter: python3
---

# Modelling yeast crosses

```{python}
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import scipy.io as sio
```

```{python}
rng = np.random.default_rng(1)
```

## Define functions

```{python}

# Parameters
num_variants = 1000  # Number of variants
num_generations = 6  # Number of generations in the intercross
num_cells_per_parent = 500  # Number of cells per parent
mean_coverage_threshold = 10  # Threshold for accurate variant calling
recombination_rate = 0.05  # Recombination rate (recombinations per unit genome length)

# Function to simulate random distribution of variants
# treats the genome as length one
def simulate_variants(num_variants):
    return np.sort(np.random.rand(num_variants))

# Function to initialize parents with marker values
def initialize_parents(num_cells_per_parent, num_variants):
    parent1 = np.random.choice([0, 1], size=num_variants)  # Parent 1 markers
    parent2 = np.random.choice([0, 1], size=num_variants)  # Parent 2 markers
    parent1 = np.zeros(num_variants)  # Parent 1 markers
    parent2 = np.ones(num_variants)  # Parent 2 markers
    cells_parent1 = np.tile(parent1, (num_cells_per_parent, 1))
    cells_parent2 = np.tile(parent2, (num_cells_per_parent, 1))
    phase_parent1 = np.ones_like(cells_parent1)
    phase_parent2 = 2*np.ones_like(cells_parent2)
    return cells_parent1, cells_parent2, phase_parent1, phase_parent2

# Function to simulate recombination with multiple segments
def simulate_recombination(cells_parent1, cells_parent2, phase_parent1, phase_parent2, num_generations, recombination_rate, variants):
    parents = np.vstack((cells_parent1, cells_parent2))
    phases = np.vstack((phase_parent1, phase_parent2))
    for generation in range(num_generations):
        print("Generation: ",generation)
        new_parents, new_phases = simulate_recombination_once(parents, phases, recombination_rate, variants)
        parents = new_parents
        phases = new_phases
    return parents, phases

# Function to simulate one round of recombination with multiple segments
# this chooses recombination sites from the set of marker locations, which isn't quite what would happen, but the effect would be the same - if the actual segment went beyond the marker any information between markers is lost anyway
# more importantly, variants are not uniformly spaced so the probability of recombination should account for that
# should pick recombination sites uniformly distributed in space and then find where the variant locations sit in relation to those points
def simulate_recombination_once(parents, phases, recombination_rate, variants):
    num_cells = parents.shape[0]
    num_variants = parents.shape[1]
    num_swaps = int(recombination_rate * num_variants)  # Number of swaps based on recombination rate
    new_cells = []
    new_phases = []
    indices = np.arange(num_cells)
    np.random.shuffle(indices)
    # print("Indices: ",indices)
    for i in range(0, len(indices), 2):
        # print("Cell pair: ",i," out of ",len(indices))
        parent1, parent2 = indices[i], indices[i+1]
        child1 = np.copy(parents[parent1])
        child2 = np.copy(parents[parent2])
        child1_phase = np.copy(phases[parent1])
        child2_phase = np.copy(phases[parent2])
        # print("child1_phase: ",child1_phase)
        # print("child2_phase: ",child2_phase)
        # choose num_swaps recombination sites
        recombination_sites = np.sort(np.random.rand(num_swaps))
        # recombination_indices = np.sort(np.random.choice(num_variants, size=num_swaps, replace=False))
        recombination_indices = variants.searchsorted(recombination_sites)
        start = recombination_indices[0]
        # swap a segment then jump two recombination sites forward
        for isite in range(0,num_swaps,2):
            if isite == num_swaps-1:
                end = num_variants
            else:
                end = recombination_indices[isite+1]
            # print("isite = ",isite,"; num_swaps = ",num_swaps)
            # print("recombination indices = ",recombination_indices,"; start = ",start,"; end = ",end)
            child1[start:end] = parents[parent2][start:end]
            child2[start:end] = parents[parent1][start:end]
            child1_phase[start:end] = phases[parent2][start:end]
            child2_phase[start:end] = phases[parent1][start:end]
            if isite < num_swaps-2:
                start = recombination_indices[isite+2]
            # print("child1_phase: ",child1_phase)
            # print("child2_phase: ",child2_phase)
        new_cells.append(child1)
        new_cells.append(child2)
        new_phases.append(child1_phase)
        new_phases.append(child2_phase)
    return np.array(new_cells), np.array(new_phases)


# Function to model missing genotypes as a Poisson process
def model_missing_genotypes(parents, mean_coverage_threshold):
    num_cells, num_variants = parents.shape
    coverage = np.random.poisson(mean_coverage_threshold, size=(num_cells, num_variants))
    missing_genotypes = np.where(coverage < mean_coverage_threshold, np.nan, parents)
    return missing_genotypes

```

```{python}
# https://stackoverflow.com/questions/1903462/how-can-i-zip-sort-parallel-numpy-arrays
def sort_by_abs_value(s, p):
    indexes = abs(s).argsort()
    return s[indexes], p[indexes]

def y_sem(y):
    return np.std(y,ddof=1)/np.sqrt(len(y))
```

## Run a simulation

```{python}
# Simulate variants
variants = simulate_variants(num_variants)

# Initialize parents with marker values
cells_parent1, cells_parent2, phase_parent1, phase_parent2 = initialize_parents(num_cells_per_parent, num_variants)

# print("Simulated Variants:", variants)
# print("Initial Parent 1 Cells:\n", cells_parent1)
# print("Initial Parent 2 Cells:\n", cells_parent2)

# Simulate recombination over generations
recombined_parents, recombined_phases = simulate_recombination(cells_parent1, cells_parent2, phase_parent1, phase_parent2, num_generations, recombination_rate, variants)

# Model missing genotypes for all parents
missing_genotypes = model_missing_genotypes(recombined_parents, mean_coverage_threshold)

# Output results
# print("Recombined Parents Markers:\n", recombined_parents)
# print("Missing Genotypes:\n", missing_genotypes)

```

## Run another simulation, storing intermediate populations

```{python}

recombination_rate = 0.2  # Recombination rate (recombinations per unit genome length)
num_generations = 6  # Number of generations in the intercross

parents = np.vstack((cells_parent1, cells_parent2))
phases = np.vstack((phase_parent1, phase_parent2))

parents_generations = []
phases_generations = []

parents_generations.append(parents)
phases_generations.append(phases)


for generation in range(num_generations):
    print("Generation: ",generation)
    new_parents, new_phases = simulate_recombination_once(parents, phases, recombination_rate, variants)
    parents = new_parents
    phases = new_phases
    parents_generations.append(parents)
    phases_generations.append(phases)


fig, ax = plt.subplots(nrows=num_generations+1,ncols=2,figsize=(8.5,7))
for generation in range(num_generations+1):
    ax[generation,0].imshow(parents_generations[generation],aspect='auto')
    ax[generation,1].imshow(phases_generations[generation],aspect='auto')

ax[0,0].set_title('Cells')
ax[0,1].set_title('Phases')
plt.show()


```

```{python}

fig, ax = plt.subplots(figsize=(8.5,7))
plt.imshow(recombined_parents,aspect='auto')
ax.set_title('Final population')

```


```{python}

# choose number of markers
num_causal_variants = 10
causal_indices = np.sort(rng.choice(num_variants, size=num_causal_variants, replace=False))
causal_weights = rng.normal(0,1,num_causal_variants)

causal_weights_sorted, causal_indices_sorted = sort_by_abs_value(causal_weights,causal_indices)

all_weights = np.zeros(num_variants)
all_weights[causal_indices] = causal_weights
```
```{python}

# generate phenotypes
phenotypes_no_noise = parents_generations[-1] @ all_weights
var_phenotypes = np.var(phenotypes_no_noise)
std_phenotypes = np.std(phenotypes_no_noise)
# unclear what is meant by "added random Gaussian noise equal to 20% of the total variance"
phenotypes = phenotypes_no_noise + rng.normal(0,0.2*std_phenotypes,phenotypes_no_noise.shape[0])
phenotypes = phenotypes_no_noise + rng.normal(0,np.sqrt(0.01*var_phenotypes),phenotypes_no_noise.shape[0])

```

```{python}
_, bins, _ = plt.hist(phenotypes_no_noise, bins=20,label='Phenotypes without noise')

plt.hist(phenotypes, bins=bins, alpha=0.5, label='with 20% additional variance')

plt.title('Histogram of phenotype distribution')
plt.legend()
```

Box plot for phenotype dependnce on strongest variant 
```{python}

i_strongest = causal_indices_sorted[-1]

p1 = phenotypes[parents_generations[-1][:,i_strongest]==1]
p0 = phenotypes[parents_generations[-1][:,i_strongest]==0]

plt.boxplot([p1,p0])
plt.title('Phenotype dependence on strongest marker')

```


```{python}
# doesn't quite work as expected - to do with rows v columns?
def vector_corr_np(data1, data2):
    data1 = np.atleast_2d(data1)
    data2 = np.atleast_2d(data2)
    mean1 = data1.mean(axis=1) 
    mean2 = data2.mean(axis=1)
    std1 = data1.std(axis=1)
    std2 = data2.std(axis=1)
    corr = ((data1*data2).mean(axis=1)-mean1*mean2)/(std1*std2)
    return corr
```

```{python}
# test = vector_corr_np(parents_generations[-1], phenotypes)
```

```{python}
# from scipy.stats import pearsonr

pr = np.zeros(num_variants)
lod1D = np.zeros(num_variants)
for i in range(num_variants): 
    pr[i] = np.corrcoef(parents_generations[-1][:,i],phenotypes)[0,1]
    # pr[i] = pearsonr(parents_generations[-1][:,i],phenotypes)
    lod1D[i] = -phenotypes.shape[0]*np.log(1-pr[i]**2)/(2*np.log(10))
```

```{python}
plt.plot(range(num_variants),pr,label='all markers')
plt.plot(causal_indices,pr[causal_indices],'xr',label='causal markers')
plt.title("Pearson's correlation for each marker with phenotype")
plt.legend()
```

```{python}
plt.plot(range(num_variants),lod1D,label='all markers')
plt.plot(causal_indices,lod1D[causal_indices],'xr',label='causal markers')
plt.title("LOD score for each marker with phenotype")
plt.legend()
```

# Working with the data

## Load trait data

```{python}
sio.matlab.matfile_version('FWselection_dependencies/trait.mat')
trait_contents = sio.matlab.loadmat('FWselection_dependencies/trait.mat')
```

`trait_contents` is a numpy array of size (1,64), each element is itself a 1x1152 numpy array of traits: 
```{python}
trait_contents['trait'].shape
trait_contents['trait'][0,0].shape
trait = np.squeeze(np.stack(np.squeeze(trait_contents['trait'])))
```

### Plot all the traits
```{python}
plt.plot(np.sort(trait).T,'*')
plt.xlabel('Strains')
plt.ylabel('Relative growth')
```

### Get the trait names

```{python}
sio.matlab.matfile_version('FWselection_dependencies/Previous Traits and Filename MAT/filename.mat')
filename_contents = sio.matlab.loadmat('FWselection_dependencies/Previous Traits and Filename MAT/filename.mat')
filename = np.squeeze(np.stack(np.squeeze(filename_contents['filename'])))
```

## Load genotype data

```{python}
sio.matlab.matfile_version('FWselection_dependencies/phasedGenotype.mat')
phasedGenotype_contents = sio.matlab.loadmat('FWselection_dependencies/phasedGenotype.mat')
phasedGenotype = phasedGenotype_contents['phasedGenotype']
```

Extract number of strains and variants (markers). 

```{python}
num_strains = phasedGenotype.shape[0]
num_variants = phasedGenotype.shape[1]

```

## Load and process `variantPos`

```{python}
sio.matlab.matfile_version('FWselection_dependencies/variantPos.mat')
variantPos_contents = sio.matlab.loadmat('FWselection_dependencies/variantPos.mat')
variantPos = np.squeeze(np.stack(np.squeeze(variantPos_contents['variantPos'])))
```


Split `variantPos` into chromosome and position variables so that we can scan 10kb on either side of each causal variant
```{python}
allPos = np.zeros(num_variants)
allChrom = np.zeros(num_variants)
pos_start = 0
for i in range(num_variants):
    tempVar = variantPos[i].split(':')
    allChrom[i] = int(tempVar[0].split('|')[1].split('_')[1])-1132
    allPos[i] = int(tempVar[1])
    if allChrom[i] == 92: 
        allChrom[i] = 17 
```

## Choose a trait to analyse

To find the index of a phenotype, e.g. Fluconazole, Tebuconazole: 

```{python}
def print_filename_indices(my_string):
    print(my_string + ' appears in indexes ')
    print(np.where(np.char.find(filename,my_string) > -1))
    print('of "filename", as')
    print(filename[np.where(np.char.find(filename,my_string) > -1)])
    print()
```

```{python}
print_filename_indices('Fluc')
print_filename_indices('Tebu')
```

```{python}

##### In figure 2B 
i_trait = 32 # '30h-Ketoconazole.txt'
# i_trait = 39 # '44h-Ketoconazole.txt'

# i_trait = 1 # '36h-Fluconazole.txt'
# i_trait = 3 # '43h-Fluconazole.txt'
# i_trait = 5 # '58h-Fluconazole.txt'

# i_trait = 33 # '30h-Tebuconazole.txt'
# i_trait = 41 # '44h-Tebuconazole.txt'
# i_trait = 48 # '72h-Tebuconazole.txt'

##### In figure 3A
# i_trait = 30 # '30h-5FU.txt'
# i_trait = 36 # '44h-5FU.txt' < peaks look wrong, big peak in Chr3
# i_trait = 0 # '36h-CuSO4.txt'
# i_trait = 8 # '74h-CuSO4.txt' < this one looks better than 36h
```

## Exclude wells that don't meet coverage threshold

```{python}
variantCalls = np.sum(abs(phasedGenotype),axis=0)
s = np.sum(abs(phasedGenotype),axis=1)

# Cutoff for number of variant calls needed to include a segregant.
rowCutoff = 5500
columnCutoff = 0

traitRow = trait[i_trait,]
# https://stackoverflow.com/questions/7820809/understanding-weird-boolean-2d-array-indexing-behavior-in-numpy
x=phasedGenotype[(s>rowCutoff) & (traitRow != -1),:][:,variantCalls>columnCutoff]
# Remove edge count and plateNorm from variant calls
# variantCalls(1:13)=[];
y=traitRow[(s>rowCutoff) & (traitRow != -1)]

```

```{python}
# Normalize trait y to mean 0, variance 1
y = y - np.mean(y)
y = y/np.std(y,ddof=1) # unbiased estimate, unlike MATLAB which does that by default
```

```{python}
pr = np.zeros(x.shape[1])
lod1D = np.zeros(x.shape[1])
for i in range(x.shape[1]): 
    pr[i] = np.corrcoef(x[:,i],y)[0,1]
    # pr[i] = pearsonr(parents_generations[-1][:,i],phenotypes)
    lod1D[i] = -x.shape[0]*np.log(1-pr[i]**2)/(2*np.log(10))

```

```{python}
i_start = 0
for i in range(1,18):
    i_chrom = np.where(allChrom==i)
    tempPos = allPos[i_chrom] + i_start
    plt.plot(tempPos,lod1D[i_chrom],label=['C' + str(i)],linewidth=0.5)
    i_start = tempPos[-1]

plt.title("LOD score with normalised phenotype " + filename[i_trait])
plt.xlabel('Marker number')
plt.ylabel('LOD score')
# plt.legend()
```

# Azole responses

> This standard analysis established a locus of strong effect on the right arm of Chromosome 4

```{python}
i_chrom = np.where(allChrom==4)
tempPos = allPos[i_chrom]
tempLOD = lod1D[i_chrom]

plt.plot(tempPos,lod1D[i_chrom],label=['C' + str(i)],linewidth=0.5)

plt.title("LOD, Chromosome 4 for " + filename[i_trait])
plt.xlabel('Marker number')
plt.ylabel('LOD score')
# plt.legend()
```

Fig 2B inset is approx 230 kb of Chromosome 4. 
```{python}
i_chrom = np.where(allChrom==4)
tempPos = allPos[i_chrom]
tempLOD = lod1D[i_chrom]
i_max = tempLOD.argmax()
i_window = np.where(abs(tempPos-tempPos[i_max]) < 10000)

plt.plot(tempPos,tempLOD,label=['C' + str(i)],linewidth=0.5)
plt.plot(tempPos[i_window],tempLOD[i_window],label=['C' + str(i)],linewidth=0.5)

plt.title("LOD, zoom on peak in Chromosome 4 for " + filename[i_trait])
plt.xlabel('Marker number')
plt.ylabel('LOD score')
plt.xlim((tempPos[i_max]-100000,tempPos[i_max]+100000))
# plt.legend()
```

The UPC2 gene (Standard Name) is also known as YDR213W (Systematic Name): [link to yeastgenome.org](https://www.yeastgenome.org/locus/S000002621). It spans locations 889751 to 892492. 

In the VCF file, YDR213W is used. Six variants are involved: 

* 889739 (c.-12C>T), 
* 889957 (c.207G>A, shows up for Clotrimazole 30h), 
* 891344 (c.1594G>A, shows up for Fluconazole 36h, Tebuconazole 30h), 
* 891466 (c.1716C>A, shows up for Prothioconazole 72h), 
* 892444 (c.2694C>T, shows up for Propiconazole 196h, Tebuconazole 44h, Voriconazole 30h), 
* 892532 (various, e.g. c.41_42delAA). 

The paper seems to consider the first five. 

```{python}
UPC2_start = 889751
UPC2_end = 892492
```

Fig 2E inset is just 20 kb, 880000-900000, showing a handful of variant locations. 

```{python}
fig, ax = plt.subplots()
ax.plot(tempPos[i_window],tempLOD[i_window],'-o',label=['C' + str(i)],linewidth=0.5)

ax.plot([(3*UPC2_start-UPC2_end)/2,(3*UPC2_end-UPC2_start)/2],[14,14],'-k')

ax.add_patch(patches.Rectangle((UPC2_start,13.5),UPC2_end-UPC2_start,1,facecolor='0.8',zorder=10))
ax.text((UPC2_start+UPC2_end)/2,14,'UPC2',horizontalalignment='center',verticalalignment='center',zorder=11)
ax.set_title("LOD, zoom on UPC2 gene on Chromosome 4 for " + filename[i_trait])
ax.set_xlabel('Marker number')
ax.set_ylabel('LOD score')
ax.set_xlim((880000,900000))
# plt.legend()
```


Searching VCF for 889739 which is the position of `-12C>T`, we find which entries in `variantPos` correspond to the five SNPs in UPC2:   
```{python}
i_12C = np.where(np.char.find(variantPos,'889739') > -1)[0][0]

i572 = i_12C+3
i2694 = i_12C+4
print(variantPos[range(i_12C,i_12C+5)])
```

Extract the two columns for those variants
```{python}
test = x[:,(i572,i2694)]
# test = phasedGenotype[:,(i572,i2694)]
```

Find the different combinations of parents across the two alleles: 'YJM/YJM', 'YJM/RM','RM/YJM','RM/RM': 

```{python}
ip1p1 = np.where((test[:,0]==1) & (test[:,1]==1))[0]
ip1m1 = np.where((test[:,0]==1) & (test[:,1]==-1))[0]
im1p1 = np.where((test[:,0]==-1) & (test[:,1]==1))[0]
im1m1 = np.where((test[:,0]==-1) & (test[:,1]==-1))[0]
plt.boxplot([y[ip1p1],y[ip1m1],y[im1p1],y[im1m1]])
plt.plot(range(1,5),[y[ip1p1].mean(),y[ip1m1].mean(),y[im1p1].mean(),y[im1m1].mean()],'*')
plt.xticks(np.arange(1,5,1),labels=['(1,1)', '(1,-1)', '(-1,1)', '(-1,-1)'])
plt.title("UPC2 variants for " + filename[i_trait])
plt.xlabel('Allele combination')
plt.ylabel('Relative growth')
plt.show()
```

Plot the mean plus or minus standard error for these four combinations: 

```{python}
y_means = [y[ip1p1].mean(),y[ip1m1].mean(),y[im1p1].mean(),y[im1m1].mean()]
y_stderror = [y_sem(y[ip1p1]),y_sem(y[ip1m1]),y_sem(y[im1p1]),y_sem(y[im1m1])]
plt.bar(range(1,5),y_means)
plt.errorbar(range(1,5),y_means,y_stderror,linestyle='None',capsize=10)
plt.xticks(np.arange(1,5,1),labels=['(1,1)', '(1,-1)', '(-1,1)', '(-1,-1)'])
plt.title("UPC2 variants for " + filename[i_trait])
plt.xlabel('Allele combination')
plt.ylabel('Relative growth (mean +/- sem)')
plt.show()

```


Check correlation of variant 
```{python}
np.corrcoef(x[:,i2694],y)
```

And compare with value from subset of data
```{python}
np.corrcoef(test[:,1],y)
```

Boxplot for max Lod on Chrom 4
```{python}
i_max = ((allChrom==4)*(lod1D)).argmax()
jp1 = np.where(x[:,i_max]==1)[0]
j0 = np.where(x[:,i_max]==0)[0]
jm1 = np.where(x[:,i_max]==-1)[0]
```

```{python}
plt.boxplot([y[jm1],y[j0],y[jp1]])
ch4_maxcor = np.corrcoef(x[:,i_max],y)[0,1]
plt.title(['Max LOD on Chr4, corr = ' + str(ch4_maxcor)])
```

# SKY1

SKY1 is also known as YMR216C. [It is on chromosome 13 between 698811 and 701039](https://www.yeastgenome.org/locus/S000004829). In the VCF file it is on lines 9079-9098 (actual part starts at 9085).

In the VCF file, YMR216C is used. 20 variants are involved, of which only two emerge from the analysis of causal variants: 

* 699044 (c.1996A>G, shows up for 5Fcytosine 44h), 
* 699680 (c.1360G>A, shows up for Rapamycin 58h)

https://hgvs-nomenclature.org/stable/background/simple/

Here's the VCF line for 699044. 

> `XIII	699044	.	T	C	48318.08	.	AC=520;AF=0.451;AN=1153;DP=38746;MQ0=0;set=Intersection;ANN=C|missense_variant|MODERATE|YMR216C|YMR216C|transcript|Transcript_YMR216C|Coding|1/1|c.1996A>G|p.Thr666Ala|1996/2229|1996/2229|666/742||	GT:AD:DP:GQ:PL	0:32,0:32:99:0,1170	0:26,0:26:99:0,900	0:22,0:22:99:0,585	0:27,0:27:99:0,945	0:29,0:29:99:0,1035	0:24,0:24:99:0,889`

`c.1996A>G` is the variant, at position 1996 in the coding DNA reference sequence. 
`p.Thr666Ala` is the predicted amino acid change at amino acid 666. This corresponds to `T666A` in the paper. 

Note that 666=1996/3 (rounded up), since three base pairs code for an amino acid. Positions 1,2,3 given AA1, 4,5,6 give AA2, etc. 

Position 699680 has 

> `XIII	699680	.	C	T	156312.08	.	AC=616;AF=0.509;AN=1210;DP=71856;MQ0=0;set=Intersection;ANN=T|missense_variant|MODERATE|YMR216C|YMR216C|transcript|Transcript_YMR216C|Coding|1/1|c.1360G>A|p.Val454Met|1360/2229|1360/2229|454/742||	GT:AD:DP:GQ:PL	1:2,92:94:99:3095,0	1:3,59:62:99:2114,0	1:0,56:56:99:2124,0	1:3,60:63:99:1883,0	1:0,88:88:99:3089,0	1:0,32:32:99:1093,0`

This has a change at amino acid 454. The other one mentioned in the paper is `D738N`, presumably should correspond to `p.Asp738Asn`. 
For confirmation of the correspondence, see [Amino Acids Representation Codes](https://www.ebi.ac.uk/pdbe/docs/roadshow_tutorial/msdtarget/AAcodes.html). 

This one is in position 698828

> `XIII	698828	.	C	T	70153.98	.	AC=573;AF=0.454;AN=1262;DP=40846;FS=0.000;MQ0=0;set=Intersection;ANN=T|missense_variant|MODERATE|YMR216C|YMR216C|transcript|Transcript_YMR216C|Coding|1/1|c.2212G>A|p.Asp738Asn|2212/2229|2212/2229|738/742||	GT:AD:DP:GQ:PL	0:19,0:19:99:0,585	0:15,0:15:99:0,405	0:19,0:19:99:0,599	0:23,0:23:99:0,675	0:19,0:19:99:0,563	0:17,0:17:99:0,469`

## Find the SKY1 variants

```{python}
i_SKY1_D738N = np.where(np.char.find(variantPos,'698828') > -1)[0][0]
i_SKY1_T666A = np.where(np.char.find(variantPos,'699044') > -1)[0][0]
```

## Gather the SKY1 5-FU trait data
```{python}

##### In figure 3A
i_trait = 30 # '30h-5FU.txt'
# i_trait = 36 # '44h-5FU.txt' < peaks look wrong, big peak in Chr3
```

## Exclude wells that don't meet coverage threshold

```{python}
variantCalls = np.sum(abs(phasedGenotype),axis=0)
s = np.sum(abs(phasedGenotype),axis=1)

# Cutoff for number of variant calls needed to include a segregant.
rowCutoff = 5500
columnCutoff = 0

traitRow = trait[i_trait,]
# https://stackoverflow.com/questions/7820809/understanding-weird-boolean-2d-array-indexing-behavior-in-numpy
x=phasedGenotype[(s>rowCutoff) & (traitRow != -1),:][:,variantCalls>columnCutoff]
# Remove edge count and plateNorm from variant calls
# variantCalls(1:13)=[];
y=traitRow[(s>rowCutoff) & (traitRow != -1)]

```

```{python}
# Normalize trait y to mean 0, variance 1
y = y - np.mean(y)
y = y/np.std(y,ddof=1) # unbiased estimate, unlike MATLAB which does that by default
```

```{python}
pr = np.zeros(x.shape[1])
lod1D = np.zeros(x.shape[1])
for i in range(x.shape[1]): 
    pr[i] = np.corrcoef(x[:,i],y)[0,1]
    # pr[i] = pearsonr(parents_generations[-1][:,i],phenotypes)
    lod1D[i] = -x.shape[0]*np.log(1-pr[i]**2)/(2*np.log(10))

```

```{python}
i_start = 0
for i in range(1,18):
    i_chrom = np.where(allChrom==i)
    tempPos = allPos[i_chrom] + i_start
    plt.plot(tempPos,lod1D[i_chrom],label=['C' + str(i)],linewidth=0.5)
    i_start = tempPos[-1]

plt.title("LOD score with normalised phenotype " + filename[i_trait])
plt.xlabel('Marker number')
plt.ylabel('LOD score')
# plt.legend()
```

## Extract the two columns for the SKY1 variants

```{python}
test = x[:,(i_SKY1_D738N,i_SKY1_T666A)]
```

Find the different combinations of parents across the two alleles: 'YJM/YJM', 'YJM/RM','RM/YJM','RM/RM': 

```{python}
ip1p1 = np.where((test[:,0]==1) & (test[:,1]==1))[0]
ip1m1 = np.where((test[:,0]==1) & (test[:,1]==-1))[0]
im1p1 = np.where((test[:,0]==-1) & (test[:,1]==1))[0]
im1m1 = np.where((test[:,0]==-1) & (test[:,1]==-1))[0]
plt.boxplot([y[ip1p1],y[ip1m1],y[im1p1],y[im1m1]])
plt.plot(range(1,5),[y[ip1p1].mean(),y[ip1m1].mean(),y[im1p1].mean(),y[im1m1].mean()],'*')
plt.xticks(np.arange(1,5,1),labels=['(1,1)', '(1,-1)', '(-1,1)', '(-1,-1)'])
plt.title("UPC2 variants for " + filename[i_trait])
plt.xlabel('Allele combination')
plt.ylabel('Relative growth')
plt.show()
```

Plot the mean plus or minus standard error for these four combinations: 

```{python}
y_means = [y[ip1p1].mean(),y[ip1m1].mean(),y[im1p1].mean(),y[im1m1].mean()]
y_stderror = [y_sem(y[ip1p1]),y_sem(y[ip1m1]),y_sem(y[im1p1]),y_sem(y[im1m1])]
plt.bar(range(1,5),y_means)
plt.errorbar(range(1,5),y_means,y_stderror,linestyle='None',capsize=10)
plt.xticks(np.arange(1,5,1),labels=['(1,1)', '(1,-1)', '(-1,1)', '(-1,-1)'])
plt.title("UPC2 variants for " + filename[i_trait])
plt.xlabel('Allele combination')
plt.ylabel('Relative growth (mean +/- sem)')
plt.show()

```


# Do it again with different genotype data

```{python}
sio.matlab.matfile_version('FWselection_dependencies/phasedGenotype_preEmptyWells.mat')
phasedGenotype_contents = sio.matlab.loadmat('FWselection_dependencies/phasedGenotype_preEmptyWells.mat')
phasedGenotype = phasedGenotype_contents['phasedGenotype']
```

```{python}
num_strains = phasedGenotype.shape[0]
num_variants = phasedGenotype.shape[1]
```
## Choose a trait to analyse

```{python}
##### In figure 2B 
i_trait = 32 # '30h-Ketoconazole.txt'
i_trait = 39 # '44h-Ketoconazole.txt'

# i_trait = 1 # '36h-Fluconazole.txt'
# i_trait = 3 # '43h-Fluconazole.txt'
# i_trait = 5 # '58h-Fluconazole.txt'

# i_trait = 33 # '30h-Tebuconazole.txt'
# i_trait = 41 # '44h-Tebuconazole.txt'
# i_trait = 48 # '72h-Tebuconazole.txt'

##### In figure 3A
# i_trait = 30 # '30h-5FU.txt'
# i_trait = 36 # '44h-5FU.txt' < peaks look wrong, big peak in Chr3
# i_trait = 0 # '36h-CuSO4.txt'
# i_trait = 8 # '74h-CuSO4.txt' < this one looks better than 36h
```

## Exclude wells that don't meet coverage threshold

```{python}
variantCalls = np.sum(abs(phasedGenotype),axis=0)
s = np.sum(abs(phasedGenotype),axis=1)

# Cutoff for number of variant calls needed to include a segregant.
rowCutoff = 5500
columnCutoff = 0

traitRow = trait[i_trait,]
# https://stackoverflow.com/questions/7820809/understanding-weird-boolean-2d-array-indexing-behavior-in-numpy
x=phasedGenotype[(s>rowCutoff) & (traitRow != -1),:][:,variantCalls>columnCutoff]
# Remove edge count and plateNorm from variant calls
# variantCalls(1:13)=[];
y=traitRow[(s>rowCutoff) & (traitRow != -1)]

```

```{python}
# Normalize trait y to mean 0, variance 1
y = y - np.mean(y)
y = y/np.std(y,ddof=1) # unbiased estimate, unlike MATLAB which does that by default
```

```{python}
pr = np.zeros(x.shape[1])
lod1D = np.zeros(x.shape[1])
for i in range(x.shape[1]): 
    pr[i] = np.corrcoef(x[:,i],y)[0,1]
    # pr[i] = pearsonr(parents_generations[-1][:,i],phenotypes)
    lod1D[i] = -x.shape[0]*np.log(1-pr[i]**2)/(2*np.log(10))

```

```{python}
i_start = 0
for i in range(1,18):
    i_chrom = np.where(allChrom==i)
    tempPos = allPos[i_chrom] + i_start
    plt.plot(tempPos,lod1D[i_chrom],label=['C' + str(i)],linewidth=0.5)
    i_start = tempPos[-1]

plt.title("LOD, pre-empty wells geno and normalised phenotype " + filename[i_trait])
plt.xlabel('Marker number')
plt.ylabel('LOD score')
# plt.legend()
```
